core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java

```diff
@@ -323,46 +324,51 @@ private String getCanonicalizedHeaderString(Map<String, List<String>> canonicali
         StringBuilder buffer = new StringBuilder();
 
         canonicalizedHeaders.forEach((headerName, headerValues) -> {
-            for (String headerValue : headerValues) {
-                appendCompactedString(buffer, headerName);
-                buffer.append(":");
-                if (headerValue != null) {
-                    appendCompactedString(buffer, headerValue);
-                }
-                buffer.append("\n");
-            }
+            buffer.append(headerName);
+            buffer.append(":");
+            buffer.append(String.join(",", trimAll(headerValues)));
+            buffer.append("\n");
         });
 
         return buffer.toString();
     }
 
     /**
-     * This method appends a string to a string builder and collapses contiguous
-     * white space is a single space.
-     *
-     * This is equivalent to:
-     *      destination.append(source.replaceAll("\\s+", " "))
+     * "The Trimall function removes excess white space before and after values,
+     * and converts sequential spaces to a single space."
+     * <p>
+     * https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+     * <p>
+     * The collapse-whitespace logic is equivalent to:
+     * <pre>
+     *     value.replaceAll("\\s+", " ")
+     * </pre>
      * but does not create a Pattern object that needs to compile the match
      * string; it also prevents us from having to make a Matcher object as well.
-     *
      */
-    private void appendCompactedString(final StringBuilder destination, final String source) {
+    private String trimAll(String value) {
         boolean previousIsWhiteSpace = false;
-        int length = source.length();
+        StringBuilder sb = new StringBuilder(value.length());
 
-        for (int i = 0; i < length; i++) {
-            char ch = source.charAt(i);
+        for (int i = 0; i < value.length(); i++) {
+            char ch = value.charAt(i);
             if (isWhiteSpace(ch)) {
                 if (previousIsWhiteSpace) {
                     continue;
                 }
-                destination.append(' ');
+                sb.append(' ');
                 previousIsWhiteSpace = true;
             } else {
-                destination.append(ch);
+                sb.append(ch);
                 previousIsWhiteSpace = false;
             }
         }
+
+        return sb.toString().trim();
+    }
+
+    private List<String> trimAll(List<String> source) {
+        return source.stream().map(this::trimAll).collect(Collectors.toList());
```

Comment:
See the other comment about doing what https://github.com/aws/aws-sdk-java-v2/pull/2502/commits/632fea82eae1e3cafad6ab8bf7028d12eb45b31e to reduce allocations

But on the topic of streams: create a JMH micro bench harness yourself but I believe stream are going to perform worse than a `for` loop here. Also, I think `Collectors.toList()` allocates the default `ArrayList` size, which may be the wrong size (either too much or too little) which requires extra instructions.